library/core/src/result.rs

```diff
@@ -1214,6 +1214,41 @@ impl<T, E> Result<Result<T, E>, E> {
     }
 }
 
+impl<T> Result<T, T> {
+    /// Returns the [`Ok`] value if `self` is `Ok`, and the [`Err`] value if
+    /// `self` is `Err`.
+    ///
+    /// In other words, this function returns the value (the `T`) of a
+    /// `Result<T, T>`, regardless of whether or not that result is `Ok` or
+    /// `Err`.
+    ///
+    /// This can be useful in conjunction with APIs such as
+    /// [`Atomic*::compare_exchange`], or [`slice::binary_search`][binary_search], but only in
```

Comment:
Binary search is the primary reason I've wanted similar, but I've also seen people want to use it in procedural macros where you use a `TokenStream` for success and failure cases.

This seems reasonable to add as an unstable feature.